ethcore/src/client/client.rs

```diff
@@ -1165,79 +1251,235 @@ impl Client {
-			BlockId::Pending => Some(self.chain.read().best_block_number() + 1),
+	fn block_number_ref(&self, id: &BlockId) -> Option<BlockNumber> {
+		match *id {
+			BlockId::Number(number) => Some(number),
+			BlockId::Hash(ref hash) => self.chain.read().block_number(hash),
+			BlockId::Earliest => Some(0),
+			BlockId::Latest => Some(self.chain.read().best_block_number()),
+		}
+	}
+}
+
+impl snapshot::DatabaseRestore for Client {
+	/// Restart the client with a new backend
+	fn restore_db(&self, new_db: &str) -> Result<(), EthcoreError> {
+		trace!(target: "snapshot", "Replacing client database with {:?}", new_db);
+
+		let _import_lock = self.importer.import_lock.lock();
+		let mut state_db = self.state_db.write();
+		let mut chain = self.chain.write();
+		let mut tracedb = self.tracedb.write();
+		self.importer.miner.clear();
+		let db = self.db.write();
+		db.restore(new_db)?;
+
+		let cache_size = state_db.cache_size();
+		*state_db = StateDB::new(journaldb::new(db.clone(), self.pruning, ::db::COL_STATE), cache_size);
+		*chain = Arc::new(BlockChain::new(self.config.blockchain.clone(), &[], db.clone()));
+		*tracedb = TraceDB::new(self.config.tracing.clone(), db.clone(), chain.clone());
+		Ok(())
+	}
+}
+
+impl Nonce for Client {
+	fn nonce(&self, address: &Address, id: BlockId) -> Option<U256> {
+		self.state_at(id).and_then(|s| s.nonce(address).ok())
+	}
+}
+
+impl Balance for Client {
+	fn balance(&self, address: &Address, state: StateOrBlock) -> Option<U256> {
+		match state {
+			StateOrBlock::State(s) => s.balance(address).ok(),
+			StateOrBlock::Block(id) => self.state_at(id).and_then(|s| s.balance(address).ok())
+		}
+	}
+}
+
+impl AccountData for Client {}
+
+impl ChainInfo for Client {
+	fn chain_info(&self) -> BlockChainInfo {
+		let mut chain_info = self.chain.read().chain_info();
+		chain_info.pending_total_difficulty = chain_info.total_difficulty + self.importer.block_queue.total_difficulty();
+		chain_info
+	}
+}
+
+impl BlockInfo for Client {
+	fn block_header(&self, id: BlockId) -> Option<::encoded::Header> {
+		let chain = self.chain.read();
+
+		Self::block_hash(&chain, id).and_then(|hash| chain.block_header_data(&hash))
+	}
+
+	fn best_block_header(&self) -> encoded::Header {
+		self.chain.read().best_block_header()
+	}
+
+	fn block(&self, id: BlockId) -> Option<encoded::Block> {
+		let chain = self.chain.read();
+
+		Self::block_hash(&chain, id).and_then(|hash| {
+			chain.block(&hash)
+		})
+	}
+
+	fn code_hash(&self, address: &Address, id: BlockId) -> Option<H256> {
+		self.state_at(id).and_then(|s| s.code_hash(address).ok())
+	}
+}
+
+impl TransactionInfo for Client {
+	fn transaction_block(&self, id: TransactionId) -> Option<H256> {
+		self.transaction_address(id).map(|addr| addr.block_hash)
+	}
+}
+
+impl BlockChainTrait for Client {}
+
+impl RegistryInfo for Client {
+	fn registry_address(&self, name: String, block: BlockId) -> Option<Address> {
+		let address = match self.registrar_address {
```

Comment:
latest version of rust supports `?` for options, so you can write:

```rust
let address = self.registrar_address?;
```

Reply:
Yeah, good point.